common/src/graph_collision.rs

```diff
@@ -12,39 +16,121 @@ use crate::{
+    chunk_queue.push_back((ChunkId::new(position.node, Vertex::A), position.local));
+
+    // Precalculate the chunk boundaries for collision purposes. If the collider goes outside these bounds,
+    // the corresponding neighboring chunk will also be used for collision checking.
+    let klein_lower_boundary = collider_radius.tanh();
+    let klein_upper_boundary =
+        ((Vertex::chunk_to_dual_factor() as f32).atanh() - collider_radius).tanh();
+
+    // Breadth-first search loop
+    while let Some((chunk, node_transform)) = chunk_queue.pop_front() {
+        let Chunk::Populated {
+                voxels: ref voxel_data,
+                ..
+            } = graph[chunk] else {
+                // Collision checking on unpopulated chunk
+                return Err(SphereCastError::OutOfBounds);
+            };
+        let local_ray = chunk.vertex.node_to_dual().cast::<f32>() * node_transform * ray;
+
+        // Check collision within a single chunk
+        let current_tanh_distance = hit.as_ref().map_or(tanh_distance, |hit| hit.tanh_distance);
+        hit = chunk_sphere_cast(
+            collider_radius,
+            voxel_data,
+            layout,
+            &local_ray,
+            current_tanh_distance,
+        )
+        .map_or(hit, |hit| {
+            Some(GraphCastHit {
+                tanh_distance: hit.tanh_distance,
+                chunk,
+                normal: math::mtranspose(&node_transform)
+                    * chunk.vertex.dual_to_node().cast()
+                    * hit.normal,
+            })
+        });
+
+        // Compute the Klein-Beltrami coordinates of the ray segment's endpoints. To check whether neighboring chunks
+        // are needed, we need to check whether the endpoints of the line segments lie outside the boundaries of the square
+        // bounded by `klein_lower_boundary` and `klein_upper_boundary`.
+        let klein_ray_start = na::Point3::from_homogeneous(local_ray.position).unwrap();
+        let klein_ray_end =
+            na::Point3::from_homogeneous(local_ray.ray_point(current_tanh_distance)).unwrap();
+
+        // Add neighboring chunks as necessary, using one coordinate at a time.
+        for axis in 0..3 {
+            // Check for neighboring nodes
+            if klein_ray_start[axis] <= klein_lower_boundary
+                || klein_ray_end[axis] <= klein_lower_boundary
+            {
+                let side = chunk.vertex.canonical_sides()[axis];
+                let next_node_transform = side.reflection().cast::<f32>() * node_transform;
+                // Crude check to ensure that the neighboring chunk's node can be in the path of the ray. For simplicity, this
+                // check treats each node as a sphere and assumes the ray is pointed directly towards its center. The check is
+                // needed because chunk generation uses this approximation, and this check is not guaranteed to pass near corners.
```

Comment:
Does this mean `BOUNDING_SPHERE_RADIUS` is incorrect? It sounds like this comment is describing a false negative case, which seems bad.

Reply:
`BOUNDING_SPHERE_RADIUS` is correct as far as I can tell. The issue is that finding the exact set of chunks subject to a sphere cast is needlessly complicated, so we use an AABB-based conservative approximation, but this approximation can include chunks outside the set of generated chunks under specific scenarios. This additional check fine-tunes the set of chunks based on a sphere-based conservative approximation instead of an AABB-based one.

Reply:
Ah, I see. Do I understand correctly that we want both because the AABB-based approximation can be much less brute force than doing the sphere-based approximation for every neighbor?

Reply:
Yeah. The sphere-based approximation alone is quite brute-force, as it doesn't depend on the direction of the ray at all. It checks whether the bounding sphere and the ray start intersects the bounding sphere of the node, and it doesn't distinguish between chunks within a given node.

The AABB approximation is nice because it checks an adjacent chunk whenever a ray gets too close to the corresponding face of the current chunk, which is quite intuitive and not very expensive to calculate.

Reply:
I adjusted the comments a bit to hopefully make this a bit less confusing for future readers of this code.